### PR TITLE
Add login, sign up, and log out functionality

### PR DESCRIPTION
**Pull request by Yacine-Saoudi**
Added front end forms and buttons to allow user to create a profile, login, and logout of their account.

- added login and signup forms that send the appropriate post requests to back end,
- login and signup each have their own components and use action creators to change state, Closes: #39 
- added a combine reducer function to allow for multiple reducers
- added a state for the login status
- added a state for user storage, Closes: #40 
- added a logout button that removes user token from the header, Closes: #42 

Closes: #11 

Signed-off-by: Yacine Saoudi <yac.sao453@gmail.com>

### CODE REVIEW

File: src/components/Login.js [hunk 1]
```diff
@@ -0,0 +1,68 @@
+/**
+ * @file Login component, takes data from form when submitted and sends a dispatch to the login reducer, then sets local storage
+ * to the token.
+ * @author Yacine Saoudi
```
Comment (OussamaSaoudi) [inline src/components/Login.js:4]: Remove Author

File: src/components/Login.js [hunk 2]
```diff
@@ -0,0 +1,68 @@
...
+ */
+import React, { useEffect } from 'react'
+import { useDispatch } from 'react-redux'
+import { login } from '../reducers/loginReducer'
+
+const Login = () => {
+  const dispatch = useDispatch()
+  /**.
```
Comment (OussamaSaoudi) [inline src/components/Login.js:12]: Remove the ```.```

File: src/components/Login.js [hunk 3]
```diff
@@ -0,0 +1,68 @@
...
+   * @method handleSubmit - takes data from the forms, cretes an obejct with it, then sends it to the login reducer to make a request.
+   *
```
Comment (OussamaSaoudi) [inline src/components/Login.js:14]: Remove the space

File: src/components/Signup.js [hunk 4]
```diff
@@ -0,0 +1,94 @@
+/**
+ * @file Signup component, user can enter their info into the fields and the component will then add their account to the database using
+ * a post request.
+ * @author Yacine Saoudi
```
Comment (OussamaSaoudi) [inline src/components/Signup.js:4]: REmove author

File: src/components/Signup.js [hunk 5]
```diff
@@ -0,0 +1,94 @@
...
+ */
+import React from 'react'
+import userService from '../services/userService'
+
+const Signup = () => {
+  /**.
```
Comment (OussamaSaoudi) [inline src/components/Signup.js:10]: Remove ```.```

File: src/components/Signup.js [hunk 6]
```diff
@@ -0,0 +1,94 @@
...
+      const signupObj = {
+        username: username,
+        email: email,
+        password: password
+      }
+      await userService.signupReq(signupObj)
+    }
+
+  }
+  return(
```
Comment (OussamaSaoudi) [inline src/components/Signup.js:43]: space after return

File: src/components/Signup.js [hunk 7]
```diff
@@ -0,0 +1,94 @@
...
+        />
+      </div>
+
+      <div>
+        password resubmit:
+        <input
+          type = 'password'
+          name = 'passwordResub'
+          onChange = {({ target }) => {
+            console.log(target.value)
```
Comment (OussamaSaoudi) [inline src/components/Signup.js:84]: Remove ```console.log```

File: src/reducers/loginReducer.js [hunk 8]
```diff
@@ -0,0 +1,48 @@
+/**
+ * @file LoginReducer contains all the reducers and action creators for the login functions.
+ * @author Yacine Saoudi
+ */
+import userService from '../services/userService'
+/**.
```
Comment (OussamaSaoudi) [inline src/reducers/loginReducer.js:6]: Remove ```.```

File: src/reducers/loginReducer.js [hunk 9]
```diff
@@ -0,0 +1,48 @@
...
+   * @reducer loginReducer - If the case is type 'LOGIN' then it takes the token received from the backend, and sets the token
+   * to the state.
+   * @param state - The state of the Reducer.
+   * @param action - The params sent from the action creator.
+   * @returns - Returns the new state as the token
+   * @param event used to prevent the default action on button press; stops the page from refreshing too soon.
+   */
+export const loginReducer = (state = {  }, action) => {
```
Comment (OussamaSaoudi) [inline src/reducers/loginReducer.js:14]: Remove double space in ```state = {  }```

File: src/reducers/loginReducer.js [hunk 10]
```diff
@@ -0,0 +1,48 @@
...
+  let newState = state
+  console.log(action.loginReq)
```
Comment (OussamaSaoudi) [inline src/reducers/loginReducer.js:16]: Remove ```console.log```

File: src/reducers/loginReducer.js [hunk 11]
```diff
@@ -0,0 +1,48 @@
...
+  switch(action.type){
+  case 'LOGIN':
+    newState = action.loginReq.token
+    userService.setToken(newState)
+    console.log(newState)
```
Comment (OussamaSaoudi) [inline src/reducers/loginReducer.js:21]: Remove ```console.log```

File: src/reducers/loginReducer.js [hunk 12]
```diff
@@ -0,0 +1,48 @@
...
+}
+/**.
+   * @method Login - Takes the object received and sends a post request to the login service where a token is received.
+   * The token is stored in the local storage, and dispatched to the reducer.
+   * @param newObj - the login info received from the frontend.
+   * @dispatch dispatches the type login and the token to the reducer.
+   * @returns
+   */
+export const login = (newObj) => {
+  console.log('LOGIN')
```
Comment (OussamaSaoudi) [inline src/reducers/loginReducer.js:35]: Remove ```console.log```

File: src/reducers/loginReducer.js [hunk 13]
```diff
@@ -0,0 +1,48 @@
...
+  case 'LOGIN':
+    newState = action.loginReq.token
+    userService.setToken(newState)
+    console.log(newState)
+    return newState
+  default:
+    return state
+  }
+}
+/**.
```
Comment (OussamaSaoudi) [inline src/reducers/loginReducer.js:27]: Remove ```.```

File: src/reducers/loginReducer.js [hunk 14]
```diff
@@ -0,0 +1,48 @@
...
+ * @author Yacine Saoudi
+ */
+import userService from '../services/userService'
+/**.
+   * @reducer loginReducer - If the case is type 'LOGIN' then it takes the token received from the backend, and sets the token
+   * to the state.
+   * @param state - The state of the Reducer.
+   * @param action - The params sent from the action creator.
+   * @returns - Returns the new state as the token
+   * @param event used to prevent the default action on button press; stops the page from refreshing too soon.
```
Comment (OussamaSaoudi) [inline src/reducers/loginReducer.js:12]: No param ```event``` present? If it is present, put it above the returns. Remove the ```-```'s pls.

File: src/services/userService.js [hunk 15]
```diff
@@ -0,0 +1,41 @@
+/**
+ * @file UserService contains all the http requests to the backend.
+ * @author Yacine Saoudi
```
Comment (OussamaSaoudi) [inline src/services/userService.js:3]: Remove author

File: src/services/userService.js [hunk 16]
```diff
@@ -0,0 +1,41 @@
...
+/**
+ * Post request to signup url.
+ *
+ * @function
+ * @param {object} signupObject - Username, password, and email that user input.
+ * @returns {object} LoginObject.
+ */
+const signupReq = async (signupObject) => {
+  const response = await axios.post(`${baseUrl}signup`, signupObject)
+  console.log(response.data)
```
Comment (OussamaSaoudi) [inline src/services/userService.js:37]: Remove ```console.log```

File: src/services/userService.js [hunk 17]
```diff
@@ -0,0 +1,41 @@
...
+ * Post request to login url.
+ *
+ * @function
+ * @param {object} loginObject - Username and password that user input.
+ * @returns {object} LoginObject.
+ */
+const loginReq = async (loginObject) => {
+  console.log(loginObject)
+  const response = await axios.post(`${baseUrl}login`, loginObject)
+  console.log(response.data)
```
Comment (OussamaSaoudi) [inline src/services/userService.js:25]: Remove ```console.log```

**Comment by OussamaSaoudi** [comment]
Mention any issues this closes as well.
Ex: 
https://github.com/orate-io/orate-io-frontend/issues/40
https://github.com/orate-io/orate-io-frontend/issues/11
https://github.com/orate-io/orate-io-frontend/issues/39

**Comment by Yacine-Saoudi** [comment]
#42 #11 #12 #38 #40 #39

**Pull request merged**